### PR TITLE
Convert Snippet and Create Snippet to Compose

### PR DESCRIPTION
**Pull request by VanyaHuamanBottleRocket**

### CODE REVIEW

**Review by BottleRocket-Colin — COMMENTED** [review]
Going to wait until other PR is merged so I'm only looking at new changes.  DM when PR#55 is merged and I'll jump on this one.

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/CreateSnippetFragmentViewModel.kt
Comment (BottleRocket-Colin) [inline app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/CreateSnippetFragmentViewModel.kt]: Technically these update functions could just be inlined inside the converter function and not clutter up view model

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetsFragment.kt
Comment (BottleRocket-Colin) [inline app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetsFragment.kt]: This UI formatting probably should not be at fragment level.  I would say best to split this logic between viewm odel and compose UI.  Reach out to me and we can discuss.

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetsFragmentViewModel.kt
Comment (BottleRocket-Colin) [inline app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetsFragmentViewModel.kt]: Why remove launchIO here?

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/SnippetsBrowser.kt
Comment (BottleRocket-Colin) [inline compose/src/main/java/com/bottlerocketstudios/compose/snippets/SnippetsBrowser.kt:47]: LazyColumn has some cool extensions that make this pattern a liltle cleaner;

LazyColumn  {
  items(list) { listItem -> composableUI...... }

}

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/SnippetsBrowser.kt
Comment (BottleRocket-Colin) [inline compose/src/main/java/com/bottlerocketstudios/compose/snippets/SnippetsBrowser.kt]: Just a nit; on single parameter function calls, you could put everything on one line.  This is just a stylistic recommendation and not required.

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/SnippetsBrowser.kt
Comment (BottleRocket-Colin) [inline compose/src/main/java/com/bottlerocketstudios/compose/snippets/SnippetsBrowser.kt:65]: Add an empty state preview

**Review by BottleRocket-Colin — COMMENTED** [review]
You're getting fast at this.  Good work, couple of minor updates and some style suggestions.

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/SnippetsBrowser.kt
Comment (VanyaHuamanBottleRocket) [reply compose/src/main/java/com/bottlerocketstudios/compose/snippets/SnippetsBrowser.kt]: Reformating

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetsFragmentViewModel.kt
Comment (VanyaHuamanBottleRocket) [reply app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetsFragmentViewModel.kt]: I must have overlooked it while I was moving things around. Fixing

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/SnippetsBrowser.kt
Comment (VanyaHuamanBottleRocket) [reply compose/src/main/java/com/bottlerocketstudios/compose/snippets/SnippetsBrowser.kt:65]: adding

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/SnippetsBrowser.kt
Comment (VanyaHuamanBottleRocket) [reply compose/src/main/java/com/bottlerocketstudios/compose/snippets/SnippetsBrowser.kt:47]: Nice

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/CreateSnippetFragmentViewModel.kt
Comment (VanyaHuamanBottleRocket) [reply app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/CreateSnippetFragmentViewModel.kt]: Changing

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetsFragment.kt
Comment (VanyaHuamanBottleRocket) [reply app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/SnippetsFragment.kt]: Removed function and moved code

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/CreateSnippetFragmentViewModel.kt
Comment (jaredstockton) [inline app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/CreateSnippetFragmentViewModel.kt:14]: @VanyaHuamanBottleRocket Thanks for catching these!

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/CreateSnippet.kt
Comment (jaredstockton) [inline compose/src/main/java/com/bottlerocketstudios/compose/snippets/CreateSnippet.kt]: @VanyaHuamanBottleRocket @BottleRocket-Colin How do we feel about passing the whole uiState instance to each UI component onscreen?

1. Is it okay to pass the full thing? This is less ideal if needing to create a preview for the composable as the data class will need to be constructed with values that aren't used by the previewed composable.
2. a) Should we try to always pass only the necessary values per component? b) If this gets fairly lengthy, should we create additional sub ui state data classes for specific composables that are subsets of the full screen ui state?
3. Should we have some wiggle room between 1 and 2? For example, if 7 of the 8 datapoints of ui state instance are needed for a composable then just pass the whole ui state object?

Initially, I'm kind of leaning towards 2a with a fallback to 3 when needed. How do you all feel about 2b vs 3?

Thoughts?

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/CreateSnippetConverter.kt
Comment (jaredstockton) [inline app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/CreateSnippetConverter.kt]: @VanyaHuamanBottleRocket Should we refactor this to `isPrivate`?

cc: @BottleRocket-Colin

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/CreateSnippetConverter.kt
Comment (jaredstockton) [inline app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/CreateSnippetConverter.kt]: @VanyaHuamanBottleRocket @BottleRocket-Colin For clarity's sake, how do you feel about `creationFailed` to give a clue to the cause of the failure (although the class name does help some with this already)?

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/CreateSnippet.kt
Comment (jaredstockton) [inline compose/src/main/java/com/bottlerocketstudios/compose/snippets/CreateSnippet.kt]: @VanyaHuamanBottleRocket Is this a common practice in compose (adjusting lineHeight based on the font size)? If so, it seems like it might be slightly more readable to use `* 1.333` instead of `* 4 / 3`

cc: @BottleRocket-Colin

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/SnippetsMockData.kt
Comment (jaredstockton) [inline compose/src/main/java/com/bottlerocketstudios/compose/snippets/SnippetsMockData.kt]: @VanyaHuamanBottleRocket Refactor this to something like `mockSnippet1` or similar

File: data/src/main/java/com/bottlerocketstudios/brarchitecture/data/model/SnippetDto.kt
Comment (jaredstockton) [inline data/src/main/java/com/bottlerocketstudios/brarchitecture/data/model/SnippetDto.kt:17]: @VanyaHuamanBottleRocket Thanks for the trailing command and extending `Dto`

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/SnippetUiModel.kt
Comment (jaredstockton) [inline compose/src/main/java/com/bottlerocketstudios/compose/snippets/SnippetUiModel.kt]: @VanyaHuamanBottleRocket To help clarify what time this represents, how about refactoring the property to be something like `formattedLastUpdatedTime`

**Review by jaredstockton — CHANGES_REQUESTED** [review]
@VanyaHuamanBottleRocket Nice work! I did leave some comments to be addressed prior to approval.

cc: @BottleRocket-Colin

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/CreateSnippetConverter.kt
Comment (VanyaHuamanBottleRocket) [reply app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/CreateSnippetConverter.kt]: Renaming

File: app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/CreateSnippetConverter.kt
Comment (VanyaHuamanBottleRocket) [reply app/src/main/java/com/bottlerocketstudios/brarchitecture/ui/snippet/CreateSnippetConverter.kt]: Renaming

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/CreateSnippet.kt
Comment (VanyaHuamanBottleRocket) [reply compose/src/main/java/com/bottlerocketstudios/compose/snippets/CreateSnippet.kt]: No this is not common. I could not find another way of putting a min number of lines that would automatically expand the view to the height of said lines. I thought of putting a height strictly based on DP but it wouldn't account for the different text size.

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/SnippetsMockData.kt
Comment (VanyaHuamanBottleRocket) [reply compose/src/main/java/com/bottlerocketstudios/compose/snippets/SnippetsMockData.kt]: renaming

File: data/src/main/java/com/bottlerocketstudios/brarchitecture/data/model/SnippetDto.kt
Comment (VanyaHuamanBottleRocket) [reply data/src/main/java/com/bottlerocketstudios/brarchitecture/data/model/SnippetDto.kt:17]: Yup

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/SnippetUiModel.kt
Comment (VanyaHuamanBottleRocket) [reply compose/src/main/java/com/bottlerocketstudios/compose/snippets/SnippetUiModel.kt]: Renaming

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/CreateSnippet.kt
Comment (VanyaHuamanBottleRocket) [reply compose/src/main/java/com/bottlerocketstudios/compose/snippets/CreateSnippet.kt]: I am ok with 2a. Let me know if I should update.

Comment (BottleRocket-Colin) [reply compose/src/main/java/com/bottlerocketstudios/compose/snippets/CreateSnippet.kt]: @br-jared-stockton / @VanyaHuamanBottleRocket - Yes, I agree.  Let's break the state down and just send separate parameters for the particular function.  That will make things more re-useable.  On that point, I also agree that if 70-80% is used, there would be a large amount of parameters, and the function in mind is single use for this screen.  Then perhaps, pass the state thru, or create a  "sub" state for the particular widget, as Jared suggested above.

But we should def lean towards our nested composables using parameters and not the state model that was already consumed above it in heirarchy.

Comment (jaredstockton) [reply compose/src/main/java/com/bottlerocketstudios/compose/snippets/CreateSnippet.kt]: @VanyaHuamanBottleRocket Let's move forward with 2a here, please sir!
@BottleRocket-Colin Agreed with your thoughts as well. Thanks for the input :)

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/CreateSnippet.kt
Comment (jaredstockton) [reply compose/src/main/java/com/bottlerocketstudios/compose/snippets/CreateSnippet.kt]: @VanyaHuamanBottleRocket If you can't find a way to set a min line count, then definitely leave a TODO comment here mentioning the reasoning for modifying line height so we can take another pass in the future :)

Comment (VanyaHuamanBottleRocket) [reply compose/src/main/java/com/bottlerocketstudios/compose/snippets/CreateSnippet.kt]: After talking with Colin I am going to remove this and just use a minDefault Height.

File: compose/src/main/java/com/bottlerocketstudios/compose/snippets/CreateSnippet.kt
Comment (VanyaHuamanBottleRocket) [reply compose/src/main/java/com/bottlerocketstudios/compose/snippets/CreateSnippet.kt]: Updating.

File: compose/src/main/java/com/bottlerocketstudios/compose/widgets/TextInputs.kt
Comment (jaredstockton) [inline compose/src/main/java/com/bottlerocketstudios/compose/widgets/TextInputs.kt:57]: @VanyaHuamanBottleRocket Thanks for the preview! Would you mind adding another one with some prefilled text?

**Review by jaredstockton — APPROVED** [review]
@VanyaHuamanBottleRocket Thanks for making the changes! Looks good (one small comment left that hopefully can make it in - if not, we can add it in another one)

File: compose/src/main/java/com/bottlerocketstudios/compose/widgets/TextInputs.kt
Comment (VanyaHuamanBottleRocket) [reply compose/src/main/java/com/bottlerocketstudios/compose/widgets/TextInputs.kt:57]: Sure. Thats easy peasy.

**Pull request merged**
